text/2019-08-21-supported-platforms.md

```diff
@@ -0,0 +1,230 @@
+        regularly tested
+      + Tested during major releases only
+      + Binaries provided when possible and functional (Eg. Releasing OS X
+        builds is not something we may always be able to do, due to new changes
+        to their architecture)
+* **Tier 3:** Official build processes do not build, test, or benchmark these
+  targets. TiKV *should probably* build on these platforms (through official or
+  ad-hoc tooling), and we'll contributor accept patches to fix or improve these
+  builds. Users are strongly discouraged from using these targets for any
+  purpose other than R&D.
+      + Example: NixOS, a Linux distro with a slightly exotic packaging concept
+      that none of our maintainers have experience with
+      + Not recommended
+      + No instructions provided
+      + No binaries/containers provided
+* **Not supported:** Testing has found the target not viable in the past. These
+  targets are either not possible to build TiKV on, or known to have serious
+  problems. Maintainers may accept patches to add or improve support if they are
+  uninvasive, but open an issue first. *Please, don't use these.*
+      + Example: Windows, we have never managed to build or successfully run
+        tests on this OS
+      + Not recommended
+      + No instructions provided
+      + No support available, bug reports about this platform may be closed
+        without reason
+
+### Target Tier/Platform Promotion
+
+Tier promotion on the **development platform** is possible via an RFC. An RFC of
+this nature should be accompanied by a series of of PRs adding content depending
+on the tier promotion. Tier promotion on the **deployment platform** is
+determined by maintainer consensus.
+
+Development Platform Promotions:
+
+* **Not supported to Tier 3:**
+      + The target can successfully run `cargo check`, `cargo clippy`, `rls`, and
+        `cargo fmt` on TiKV without patches or modifications
+      + The target can successfully run `cargo build` and `cargo build --release`
+        on TiKV without patches or modifications
+      + The target can produce development builds of TiKV, they do not
+        necessarily run
+* **Tier 3 to Tier 2:**
+      + The target can host and run TiKV as a service, and serve requests
+      + The target can actively particpate in a TiKV cluster containing TiKV
+        nodes running on tier 1 targets
+      + The target can pass TiKV's test suite and benchmark suite
+      + The proposal is submitted by a regular contributor to TiKV, and they
+        state they intend to maintain support
+* **Tier 2 to Tier 1:**
+      + The target has reliably passed TiKVs test and benchmark suites for a
+        period of 3 months
+      + The target can use an interactive debugger on TiKV and documentation is present
+      + The target supports at least one fuzzer and fuzzing has been performed
+      + The target is used as a development machine by at least one maintainer
+        or three regular contributors
```

Comment:
These all sound like deployment constraints rather than dev ones

Reply:
I would definitely not like to see a development platform reach "Officially Supported" status without all of these present.

Reply:
I don't think any of the dev platforms in this RFC (except CentOS) satisfy all these constraints (fuzzing in particular seems a big ask, I would expect us to only fuzz one platform)

Reply:
Fuzzing works on all Linuxes listed just fine and last I checked it was possible to fuzz on mac as well.

Reply:
But we certainly can't run all tests and benches on OS X. Has anybody ever tried running OS X and Linux in the same cluster? (and these are all just tier 2 things).

"The target has reliably passed TiKVs test and benchmark suites for a period of 3 months" - given the number of flaky tests, I would argue no platform meets this requirement.

What I meant about fuzzing is that I don't think it was necessary. Hypothetically, if Windows satisfied all the other requirements, but couldn't fuzz, I'd be happy for it to be a supported dev platform because we can fuzz elsewhere.

In other words, I don't think you need to run TiKV to develop it, you just need to be able to build for a platform where you can run it (by analogy, in mobile development, there are deployment requirements for phones, but you are not expected to develop on one, and the machine you use to develop on does not need to run phone apps).